### DISCUSSION TITLE
Leaderboard Display Consolidation

### DISCUSSION THREAD

**Discussion by Etron-RA** [RetroAchievements/RAIntegration#880] [Completed]
When multiple leaderboards are active at once it can create a very busy display of the leaderboard values updating.  I've noticed that some leaderboards share the same value and could be consolidated into one value.  For example, I've created a few speedrun leaderboards in [Pitfall!](https://retroachievements.org/game/11191) of varying lengths that all start at the beginning of the game and end when certain treasure(s) are collected.  So the time value is the same for each speedrun leaderboard.  The same is true for the two score leaderboards, one is for overall score and the other is a deathless score.

![Multiple Indentical Leaderboard Values](https://i.imgur.com/L9lYoQs.png)

Ideally, I think the values on the leaderboard in the above screenshot could be consolidated from:
```
005812
005812
3
1:18.98
1:18.98
1:18.98
```
to:
```
005812
3
1:18.98
```
without losing any information.

**Comment by Tsearo** [comment]
Perhaps another flag can be added to the leaderboard definitions to denote not displaying it in the emulator. That way the dev can choose to hide the additional leaderboards that may submit/cancel earlier while keeping the longest one that would stay up that also would share the same value.

**Reply by Etron-RA** [reply]
A flag might be a better solution, that way the it doesn't consolidate unrelated leaderboards that happen to be the same value for a few frames.  It would work for the above example if the two shorter speedruns and the deathless score were hidden.

**Comment by Etron-RA** [comment]
To follow up on this idea I saw a leaderboard technique used in [-Homebrew- VB Racing (Virtual Boy)](https://retroachievements.org/game/13164) by @kmpers that had an overall speedrun displayed and smaller speedruns for each track which were not displayed.  I applied the same idea to Pitfall! and consolidated the score and speedrun leaderboards.  In the process I was able to add three more speedruns without adding to the visual clutter.

For the speedrun leaderboards the beat game speedrun starts at the start of the game, cancels on game over, and submits when you collect all of the treasures.  This means for the entire game the time is displayed at the bottom of the screen.  The other speedruns only have a start conditions, an always false cancel, and an always true submit.  Using this technique the leaderboard starts/submits in the same frame. For example, the fastest time to get a gold bar leaderboard starts/submits when the players grabs the first gold bar.  To prevent the leaderboard from submitting everytime the play picks up a gold bar a hit is used to lock the start condition, the hit is then reset by a game over.

Similarly, the overall high score leaderboard starts at the start of the game, cancels on an impossible score, and submits on game over or on reset.  This means for the entire game the score is displayed at the bottom of the screen.  The deathless score leaderboard starts/submits when the player dies for the first time.  

There are a couple of issues with this technique: the user isn't notified that the leaderboard active and you can't tally hits in the measure since it's only active for a frame.  Ideally, we'd have a do not display flag like @Tsearo suggested above. In the meanwhile, this was a pretty easy fix that really improves the play experince.  I hope that others find it useful too.

If anyone is interested in the details, the [source](https://github.com/Etron-RA/RetroAchievements/blob/main/Backup%20Code/Pitfall!%20(A2600)/Pitfall!%20Leaderboards.rascript) is on github.

**Comment by Jamiras** [comment]
Implemented in #976
